include/swift/AST/NameLookup.h

```diff
@@ -536,8 +575,91 @@ class FindLocalVal : public StmtVisitor<FindLocalVal> {
   void visitCatchStmt(CatchStmt *S);
   
 };
+  
+  
+/// The bridge between the legacy UnqualifedLookupFactory and the new ASTScope
+/// lookup system
+class AbstractASTScopeDeclConsumer {
+public:
+  AbstractASTScopeDeclConsumer() {}
+
+  virtual ~AbstractASTScopeDeclConsumer() = default;
+
+  /// Called for every ValueDecl visible from the lookup.
+  /// Returns true if the lookup can be stopped at this point.
+  /// BaseDC is per legacy
+  /// Takes an array in order to batch the consumption before setting
+  /// IndexOfFirstOuterResult when necessary.
+  virtual bool consume(ArrayRef<ValueDecl *> values, DeclVisibilityKind vis,
+                       NullablePtr<DeclContext> baseDC = nullptr) = 0;
+
+  /// Eventually this functionality should move into ASTScopeLookup
+  virtual std::pair<bool, Optional<bool>>
+  lookupInSelfType(NullablePtr<DeclContext> selfDC, DeclContext *const scopeDC,
+                   NominalTypeDecl *const nominal,
+                   Optional<bool> isCascadingUse) = 0;
+
+  virtual void stopForDebuggingIfTargetLookup() = 0;
+};
+  
+/// Just used to print
+/// Used to gather lookup results
+class ASTScopeDeclGatherer : public AbstractASTScopeDeclConsumer {
+  SmallVector<ValueDecl *, 32> values;
+
+public:
+  virtual ~ASTScopeDeclGatherer() = default;
+
+  bool consume(ArrayRef<ValueDecl *> values, DeclVisibilityKind vis,
+               NullablePtr<DeclContext> baseDC = nullptr) override;
 
+  /// Eventually this functionality should move into ASTScopeLookup
+  std::pair<bool, Optional<bool>>
+  lookupInSelfType(NullablePtr<DeclContext>, DeclContext *const,
+                   NominalTypeDecl *const,
+                   Optional<bool> isCascadingUse) override {
+    return std::make_pair(false, isCascadingUse);
+  }
+
+  void stopForDebuggingIfTargetLookup() override {}
+
+  ArrayRef<ValueDecl *> getDecls() { return values; }
+};
 } // end namespace namelookup
+
+/// The interface into the ASTScope subsystem
+class ASTScope {
+  friend class ast_scope::ASTScopeImpl;
+  ast_scope::ASTSourceFileScope *const impl;
+
+public:
+  ASTScope(ast_scope::ASTSourceFileScope *sfs) : impl(sfs) {}
+  static ASTScope *createScopeTreeFor(SourceFile *);
+  static Optional<bool>
+  unqualifiedLookup(SourceFile *, DeclName, SourceLoc,
+                    const DeclContext *startingContext,
+                    Optional<bool> isCascadingUse,
+                    namelookup::AbstractASTScopeDeclConsumer &);
+
+  void dump() const;
```

Comment:
Can you make sure that the dump methods are only in assert builds? That is what we generally do.

Reply:
I understand the desire to slim down the compiler and sometimes `dump` is benind `ifndef NDBUG`. But also I see places where `void() dump` seems to appear all the time, and I worry about losing that debugging facility in release builds. In the code of Doug's that I started with, he used: `LLVM_ATTRIBUTE_DEPRECATED` and I'll do that for my `dump`.